### PR TITLE
fix: Account for offsets when creating new Position instances

### PR DESCRIPTION
**Pull request by DylanPiercey**
Another followup to #16936 and #16849...

* The column passed for `new Position` in the `curPosition` api is incorrectly subtracting a "source position" (`index`) from an "offset position" (`this.lineStart`). This causes the column to be incorrect. Since `lineStart` is a "source position" we can subtract it directly from the `pos`.
* I searched through the codebase for all `new Position` calls to ensure that correctly offset line/column/index was passed and discovered the `index` for `firstInvalidTemplateEscapePos` was not correctly offset.

As I mentioned in the other PR I am attempting to use these new apis in [Marko](https://github.com/marko-js/marko) and with these two (hopefully final) tweaks the Marko test suite now passes. There's a fair amount of "source position" testing and snapshots in that suite so hopefully a good sign that we're not missing anything else.

Note I added another test that is a copy of all of the "start" position override tests just so that I could manually diff/compare all of the test suites from that `output.json` to see that all the offsetting looks correct. I think it's helpful to keep that test there to reference.

Again I thank all of you for working with me on this 🙏 

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          | 👍
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

**Comment by liuxingbaoyu** [comment]
Well, the FUZZ test is not running completely correctly, I will investigate.

**Comment by nicolo-ribaudo** [comment]
Would you like to open a PR adding Marko to out e2e test suite?

https://github.com/babel/babel/blob/main/.github/workflows/e2e-tests.yml

> Well, the FUZZ test is not running completely correctly, I will investigate.

It's that we are ignoring locations in the fuzz test so we cannot catch this, right?

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by liuxingbaoyu** [comment]
> It's that we are ignoring locations in the fuzz test so we cannot catch this, right?

Yes!

**Comment by nicolo-ribaudo** [comment]
Maybe the adjust function in those tests would adjust the locations in the AST so that we don't need to ignore them?

It's been very long since I looked at that code, I don't remember well how it works.

**Comment by liuxingbaoyu** [comment]
> Maybe the adjust function in those tests would adjust the locations in the AST so that we don't need to ignore them?
> 
> It's been very long since I looked at that code, I don't remember well how it works.

Amazingly correct! It's just a little more complicated to fix since we automatically adjust options in the parser.
Also, fuzz testing found a few bugs that I'll push as soon as possible.

**Comment by nicolo-ribaudo** [comment]
Really happy with how the tests turned out :)

File: packages/babel-parser/test/helpers/to-fuzzed-options.js [hunk 1]
```diff
@@ -66,8 +73,6 @@ export default function toFuzzedOptions(options) {
     [randomIndex, 1, 0],
     [randomIndex, randomLine, randomColumn],
     [false, randomLine, false],
-    [false, false, randomColumn],
-    [randomIndex, false, false],
```
Comment (JLHwung) [inline packages/babel-parser/test/helpers/to-fuzzed-options.js:70]: Can we add `[randomIndex, false, randomIndex]` here? It seems that this option does not overlap with others.

**Review by JLHwung — APPROVED** [review]
Well done!

File: packages/babel-parser/test/helpers/to-fuzzed-options.js (hunk 1 above)
Comment (liuxingbaoyu) [reply packages/babel-parser/test/helpers/to-fuzzed-options.js:70]: Great idea!

**Comment by nicolo-ribaudo** [comment]
Released

**Pull request merged**